### PR TITLE
[WIP] Adding onError Func to the transport seam

### PR DESCRIPTION
**Pull request by andreasohlund**
Related to #3792

### CODE REVIEW

**Comment by andreasohlund** [comment]
@timbussmann @tmasternak I've now cleaned up the native receiver:

https://github.com/Particular/NServiceBus/pull/3879/files#diff-9639a000ccc096beae02e6874ab91d82

what do you think?

File: src/NServiceBus.Core/StartableEndpoint.cs [hunk 1]
```diff
@@ -108,18 +108,33 @@ IEnumerable<TransportReceiver> BuildPipelines(IPipelineCache cache)
...
             foreach (var satellitePipeline in settings.Get<SatelliteDefinitions>().Definitions)
             {
                 var satellitePushSettings = new PushSettings(satellitePipeline.ReceiveAddress, errorQueue, purgeOnStartup, satellitePipeline.RequiredTransportTransactionMode);
-                yield return new TransportReceiver(satellitePipeline.Name, builder, satellitePushSettings, dequeueLimitations, satellitePipeline.OnMessage);
+                yield return new TransportReceiver(satellitePipeline.Name,
+                    builder,
+                    satellitePushSettings,
+                    dequeueLimitations,
+                    satellitePipeline.OnMessage,
+                    (b, context) => InvokeSatelliteError(b, context));
```
Comment (timbussmann) [inline src/NServiceBus.Core/StartableEndpoint.cs]: should that also be just `InvokeSatelliteError`? If I remember correctly from one of daniel's tests, the normal lambda with the method invocation inside was faster though?

File: src/NServiceBus.Core/Transports/Msmq/ReceiveStrategy.cs [hunk 2]
```diff
@@ -1,13 +1,140 @@
...
+            ErrorQueue.Send(message, transactionType);
+        }
+
+        protected async Task<bool> TryProcessMessage(Message message, Dictionary<string, string> headers, TransportTransaction transaction)
+        {
+            using (var tokenSource = new CancellationTokenSource())
+            {
+                using (var bodyStream = message.BodyStream)
+                {
+                    var pushContext = new MessageContext(message.Id, headers, bodyStream, transaction, tokenSource, new ContextBag());
```
Comment (timbussmann) [inline src/NServiceBus.Core/Transports/Msmq/ReceiveStrategy.cs]: rename variable to messageContext?

Comment (andreasohlund) [inline src/NServiceBus.Core/Transports/Msmq/ReceiveStrategy.cs]: fixed

File: src/NServiceBus.Core/Transports/Msmq/ReceiveWithNativeTransaction.cs [hunk 3]
```diff
@@ -3,68 +3,92 @@ namespace NServiceBus
...
+                        MsmqFailureInfoStorage.ProcessingFailureInfo failureInfo;
 
-                        await onMessage(pushContext).ConfigureAwait(false);
-                    }
+                        var shouldTryProcessMessage = true;
 
-                    if (cancellationTokenSource.Token.IsCancellationRequested)
+                        if (failureInfoStorage.TryGetFailureInfoForMessage(message.Id, out failureInfo))
+                        {
+                            var shouldRetryImmediately = await HandleError(message, headers, failureInfo.Exception).ConfigureAwait(false);
```
Comment (andreasohlund) [inline src/NServiceBus.Core/Transports/Msmq/ReceiveWithNativeTransaction.cs]: Not sure what you mean by pass up?

**Comment by tmasternak** [comment]
This has been substituted by https://github.com/Particular/NServiceBus/pull/3892

**Pull request closed without merging**